File: src/bin/1143-longest-common-subsequence.py
# @leetup=info id=1143 lang=python3 slug=longest-common-subsequence



# @leetup=code
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        # N1 = len(text1)
        # N2 = len(text2)
        # dp = [[0] * (N2 + 1) for _ in range(N1 + 1)]
        # for i in range(1, N1 + 1):
        #     for j in range(1, N2 + 1):
        #         dp[i][j] = max(
        #             dp[i - 1][j],
        #             dp[i][j - 1],
        #             dp[i - 1][j - 1] + (1 if text1[i - 1] == text2[j - 1] else 0),
        #         )

        dp = [[0] * (len(text2) + 1) for _ in range(len(text1) + 1)]
        for i, a in enumerate(text1):
            for j, b in enumerate(text2):
                if a == b:
                    dp[i + 1][j + 1] = dp[i][j] + 1
                else:
                    dp[i + 1][j + 1] = max(dp[i + 1][j], dp[i][j + 1])

        return dp[-1][-1]


# @leetup=code

File: src/bin/1143-longest-common-subsequence.rs
// @leetup=info id=1143 lang=rust slug=longest-common-subsequence

// @leetup=code
impl Solution {
    pub fn longest_common_subsequence(text1: String, text2: String) -> i32 {
        let mut dp = vec![vec![0; text2.len() + 1]; text1.len() + 1];
        for (i, a) in text1.chars().enumerate() {
            for (j, b) in text2.chars().enumerate() {
                dp[i + 1][j + 1] = if a == b {
                    dp[i][j] + 1
                } else {
                    dp[i + 1][j].max(dp[i][j + 1])
                }
            }
        }
        dp[text1.len()][text2.len()]
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::longest_common_subsequence;
}
